src/app/modules/item/helpers/item-permissions-constraints.ts

```diff
@@ -0,0 +1,242 @@
+
+  // For all canWatch except 'none'
+  if (!atLeast(canViewValues, receiverPermissions.canView, 'content')) {
+    errors.push(`${thisUserNeedsText} <b>${permissionsInfoString.canView.string}</b> ${toBeAtLeastText} <b>${
+      permissionsInfoString.canView.content}</b>`);
+  }
+
+  if (receiverPermissions.canWatch === 'answer_with_grant' && !giverPermissions.isOwner) {
+    errors.push(youNeedToBeOwnerText);
+  } else {
+
+    // if receiverPermissions.canWatch is 'result' or 'answer'
+    if (!(giverPermissions.canWatch === 'answer_with_grant')) {
+      errors.push(`${youNeedText} <b>${permissionsInfoString.canWatch.string}</b> ${toBeText} <b>${
+        permissionsInfoString.canWatch.answer_with_grant}</b>`);
+    }
+  }
+
+  return { canWatch: errors.length === 0 ? undefined : errors };
+}
+
+function validateCanEdit(receiverPermissions: GroupPermissions, giverPermissions: PermissionsInfo): { canEdit?: string[] } {
+
+  if (receiverPermissions.canEdit === 'none') return {};
+
+  const errors: string[] = [];
+
+  // For all can_edit except 'none'
+  if (!atLeast(canViewValues, receiverPermissions.canView, 'content')) {
+    errors.push(`${thisUserNeedsText} <b>${permissionsInfoString.canView.string}</b> ${toBeAtLeastText} <b>${
+      permissionsInfoString.canView.content}</b>`);
+  }
+
+  if (receiverPermissions.canEdit === 'all_with_grant' && !giverPermissions.isOwner) {
+    errors.push(youNeedToBeOwnerText);
+  } else {
+
+    // if receiverPermissions.canEdit is 'children' or 'all_with_grant'
+    if (!(giverPermissions.canEdit === 'all_with_grant')) {
+      errors.push(`${youNeedText} <b>${permissionsInfoString.canEdit.string}</b> ${toBeText} <b>${
+        permissionsInfoString.canEdit.all_with_grant}</b>`);
+    }
+  }
+
+  return { canEdit: errors.length === 0 ? undefined : errors };
+}
+
+function validateCanMakeSessionOfficial(
+  previousValue: boolean,
+  receiverPermissions: GroupPermissions,
+  giverPermissions: PermissionsInfo
+): { canMakeSessionOfficial?: string[] } {
+  const errors: string[] = [];
+
+  if (receiverPermissions.canMakeSessionOfficial && !previousValue) {
+    if (!giverPermissions.isOwner) {
+      errors.push(youNeedToBeOwnerText);
+    }
+    if (!atLeast(canViewValues, receiverPermissions.canView, 'content')) {
+      errors.push(`${thisUserNeedsText} <b>${permissionsInfoString.canView.string}</b> ${toBeAtLeastText} <b>${
+        permissionsInfoString.canView.content}</b>`);
+    }
+  }
+
+  return errors.length === 0 ? {} : { canMakeSessionOfficial: errors };
+}
+
+function validateIsOwner(
+  previousValue: boolean,
+  receiverPermissions: GroupPermissions,
+  giverPermissions: PermissionsInfo
+): { isOwner?: string[] } {
+
+  if (receiverPermissions.isOwner && !previousValue && !giverPermissions.isOwner) {
+    return { isOwner: [ youNeedToBeOwnerText ] };
+  }
+  return {};
+}
+
+export function permissionsConstraintsValidator(
+  receiverPermissions: GroupPermissions,
```

Comment:
This is to handle the change for `is_owner` and `can_make_session_official`.

In the following explication, the giver is the user changing the permissions, and the receiver is the user whose permissions are being changed.
To my understanding, the receiver being an owner of the item (permission maybe granted by someone else) doesn't necessarily mean the giver is one, so the form would be invalidated right away, because the current giver doesn't have the permission to grant `is_owner` to the receiver.

That's why I need to have the initial value of the form, so there is an exception (the aforementioned case, receiver is already an owner, but not the giver) and only test the case "receiver goes from not-owner to owner"

I'm not 100% sure of this edge case, but that's what I deducted from the permissions constraints in the [devdoc](https://france-ioi.github.io/algorea-devdoc/design/access-rights/items/#changing-and-propagating-permissions) for the "is_owner=true" granted perm.